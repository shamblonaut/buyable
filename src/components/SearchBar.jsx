import styled from "styled-components";
import { Search as LucideSearch } from "lucide-react";

const Bar = styled.form`
  display: flex;
  align-items: center;
  margin: 16px;
  border: 1px solid grey;
  border-radius: 4px;
  height: min-content;
`;

const SearchIcon = styled(LucideSearch)`
  margin: 8px;
`;

const Input = styled.input`
  appearance: none;
  border: none;
  outline: none;

  flex: 1;
  padding: 8px;
  border-radius: inherit;
`;

const SearchBar = ({ prompt, query, setQuery }) => {
  const handleSearchInput = (event) => setQuery(event.target.value);

  return (
    <Bar>
      <SearchIcon />
      <Input
        type="text"
        placeholder={prompt}
        value={query}
        onChange={handleSearchInput}
      />
    </Bar>
  );
};

export default SearchBar;
