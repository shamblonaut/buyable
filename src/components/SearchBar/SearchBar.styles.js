import styled from "styled-components";
import { LucideSearch } from "lucide-react";

export const Bar = styled.form`
  display: flex;
  align-items: center;
  margin: 16px;
  border: 1px solid grey;
  width: max-content;
  height: min-content;
  min-width: 80%;

  background-color: var(--color-light);
  border: 3px solid var(--color-dark);
  border-radius: 4px;
`;

export const SearchIcon = styled(LucideSearch)`
  margin: 8px;
`;

export const Input = styled.input`
  appearance: none;
  border: none;
  outline: none;
  background: none;
  color: inherit;

  flex: 1;
  padding: 8px;
  border-radius: inherit;
`;
