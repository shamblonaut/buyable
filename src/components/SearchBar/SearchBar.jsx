import PropTypes from "prop-types";

import { Bar, SearchIcon, Input } from "./SearchBar.styles";

const SearchBar = ({ prompt, query, setQuery }) => {
  const handleSearchInput = (event) => setQuery(event.target.value);

  return (
    <Bar>
      <SearchIcon />
      <Input
        type="text"
        placeholder={prompt ?? ""}
        value={query}
        onChange={handleSearchInput}
      />
    </Bar>
  );
};

SearchBar.propTypes = {
  prompt: PropTypes.string,
  query: PropTypes.string.isRequired,
  setQuery: PropTypes.func.isRequired,
};

export default SearchBar;
