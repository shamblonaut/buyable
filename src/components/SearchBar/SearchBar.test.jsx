import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import SearchBar from "./SearchBar";

const mockPrompt = "Mock Search Prompt";
const mockQuery = "Mock Query";
const mockSetQuery = vi.fn();

describe("SearchBar component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(
      <SearchBar
        prompt={mockPrompt}
        query={mockQuery}
        setQuery={mockSetQuery}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("handles input properly", async () => {
    const user = userEvent.setup();

    let query = "";
    let setQuery = vi.spyOn(
      {
        setQuery: (newQuery) => {
          query = newQuery;
        },
      },
      "setQuery",
    );

    const component = (
      <SearchBar prompt={mockPrompt} query={query} setQuery={setQuery} />
    );

    const { rerender } = render(component);
    const searchInput = screen.getByRole("textbox", {
      name: "product-search-input",
    });

    expect(searchInput).toHaveValue("");

    await user.type(searchInput, "test query");
    rerender(component);

    waitFor(() => {
      expect(searchInput).toHaveValue("test query");
    });
  });
});
