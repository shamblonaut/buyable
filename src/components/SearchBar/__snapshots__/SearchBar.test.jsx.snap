// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar component > renders correctly 1`] = `
<div>
  <form
    class="sc-bRKDuR jvqWos"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-search sc-hvigdm cNdWqb"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m21 21-4.34-4.34"
      />
      <circle
        cx="11"
        cy="11"
        r="8"
      />
    </svg>
    <input
      aria-label="product-search-input"
      class="sc-fhHczv jgMaqb"
      placeholder="Mock Search Prompt"
      type="text"
      value="Mock Query"
    />
  </form>
</div>
`;
